service/history/historyEngine.go

```diff
@@ -404,6 +361,71 @@ func (e *historyEngineImpl) registerNamespaceFailoverCallback() {
 	)
 }
 
+func (e *historyEngineImpl) listenToClusterMetadataChange() {
+	e.clusterMetadata.RegisterMetadataChangeCallback(
+		fmt.Sprintf(historyServiceCallbackID, e.shard.GetShardID()),
+		func(oldClusterMetadata map[string]*cluster.ClusterInformation, newClusterMetadata map[string]*cluster.ClusterInformation) {
+			e.replicationTaskProcessorsLock.Lock()
+			defer e.replicationTaskProcessorsLock.Unlock()
+
+			for clusterName := range newClusterMetadata {
+				if clusterName == e.currentClusterName {
+					continue
+				}
+				if processor, ok := e.replicationTaskProcessors[clusterName]; ok {
+					processor.Stop()
+					delete(e.replicationTaskProcessors, clusterName)
+				}
```

Comment:
I don't follow the logic here. You need to check if the existing remote cluster in replicationTaskProcessors no longer present in newClusterMetadata, then you stop and delete it.

Reply:
The metadata register update the metadata when the fields are updated 1. enabled, 2. RPCAddress 3. Initial failover version. 
Whenever we received a callback there are three cases 1. remote remote cluster 2. add new cluster. 3. refresh remote cluster. This check covers the case 1 and case 3. 
Also add some comments about it